### ISSUE TITLE
BUg with Backorder processing and warnings icw MSI

### ISSUE THREAD

**Issue by seansan** [magento/magento2#28824]
### Preconditions (*)
1. M2.3.5, php72
2. Enable "Allow Qty Below 0 and Notify Customer" using default config in system settings
3. create 1 simple product with qty = 1 for MSI source and 'Out-of-Stock Threshold' = - 5 
4. check in grid, salable qty should now be 6

### Steps to reproduce (*)
1. Goto frontend and order qty 1, success, stay on product page
2. Order another qty =1 (2 total)
_Unexpected 1:_ there is no warning at all about backorders a) not as a system message and b) not in the minicart!!
3. Goto cart, here you see the msg under the product "We don't have as many quantity as you requested, but we'll back order the remaining 1."
4. Go back to the product, order another qty = 4 (6 total), success, stay on product page
_Unexpected 1_ again
5. Now order another qty 2 (8 total)
_Unexpected 2_: Now 2 error messages appear. First the message "We don't have as many quantity as you requested, but we'll back order the remaining 5." belonging to he salable qty. Only: we would expect to see this message after step 2 qand after step 4 ... not now! This would solve _unexpected 1_
_Unexpected 3_: Also a 2nd error message appears showing "The requested qty is not available", her I would expect to read "The requested qty **of 2** is not available". The qty should be shown in the message.

### Expected result (*)
1. Unexpected 1: there is no warning at all after step 2 and 4
2. Unexpected 2: after step 5 we dont expect the warning msg "We don't have as many ... etc"
3. Unexpected 3: after step 5 if the qty is really not available (qty > salable qty) then show a msg where we can see the qty that is over the salable qty : in this case 2 are not available

In addition to Unexpected 3. We dont know if the available qty should be added, then the delta between salable and qty should be reported (so partially added to cart, what ever part of the qty is available). Or if the qty cannot be added as a whole amount then we deny the total qty and report this.

### Actual result (*)

![image](https://user-images.githubusercontent.com/652395/85283450-afbcdd00-b48d-11ea-92e6-0b49005f00de.png)

### Severity
- [X ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._

### Secondary problems to test (*)
- Setup local stock qith QTY, set 'Out-of-Stock Threshold' = - 5 ... salable qty should be 5. Add 5 to cart
- after this add 1 more to cart, this should not be possible ....and a warning is expected ... but magento just reloads the page there is no warning

**Comment by swathishs** [comment]
Hi @seansan , All issue specific for MSI should be reported in the MSI repo

fyi: @sidolov @smoskaluk @sdzhepa

It seems this issue MSI specific and should be fixed in the MSI repository.

EngCom team, please confirm and transfer this issue to the https://github.com/magento/inventory/issues

Closing this issue

**Comment by seansan** [comment]
Is this MSI? I would expect backorder logic ... 

!!!! As long as this issue does not gets lost, lost in translation or overly discussed (in the end the users need a fix ;)

**Comment by swathishs** [comment]
@seansan: Since the issue can be recreated by enabling the MSI module right?. Don't worry the issue will be transferred to the MSI repository

**Comment by seansan** [comment]
maybe to occurs also without MSI ? (this I cannot test)

**Comment by swathishs** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @swathishs. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @swathishs, here is your [Magento instance](https://i-28824-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-28824-2-4-develop.instances.magento-community.engineering/admin_4a8d](https://i-28824-2-4-develop.instances.magento-community.engineering/admin_4a8d)
Login: `e0ab881a` Password: `d7c03da39c9c`
Instance will be terminated in up to 3 hours.

**Comment by seansan** [comment]
**Error 1** confirmed on vanilla (qty = 3, Out-of-Stock Threshold = -5, salable qty =8)

1. add 3 to cart, success
2. add 1 to cart (total = 4) : expected warning on screen + on minicart "We don't have as many quantity as you requested, but we'll back order the remaining 1."

**Error 2** confirmed on vanilla (qty = 3, Out-of-Stock Threshold = -5, salable qty =8)
1. add 8 to cart, success (but with error 1 missing warning)
2. add 1 more to cart = 9 should not be possible : expected error cannot add to cart, but instead nothing

**Comment by swathishs** [comment]
@seansan : You can check-in 2.4 develop an instance  above
Where should we set salable quantity in admin? 
Then we can't set Out-of-Stock Threshold = -5 (Negative value)
It will be more clear by putting the screenshot

**Comment by gwharton** [comment]
Fixed in 2.4-develop

**Comment by seansan** [comment]
@gwharton are you sure?

There are three different warnings. And they should display on 3 locations. 

"**There are 3 kinds of messages to be displayed in the frontend. 

The messages need to be visible on each! page: the productpage, the minicart and the cart. These are:

1. ""Only [salable qty] left"" should be displayed when qty reaches the minimal qty.

2. ""The requested qty is not available"" should be displayed when ordering more items than salable qty

3. ""We don't have as many quantity as you requested, but we'll back order the remaining 3"" should be displayed when ordering items that need to be backordered"I 


In addition. These three warning should be shown everywhere : in Cart. On product add (msg). And in mini cart.

**Comment by gwharton** [comment]
1. I overrode this functionality in my store to display how many physical I have in stock on the product page. So if I have 5 in stock, and the threshold is -20, my salable qty is 25, but I still output a message on the frontend saying only 5 left. Once the salable quantity gets down to 20, i.e 0 in stock and more orders would be backordered, then I change the button to say "Backorder" instead of "Add to cart" and update the "Only x left" message to say "Available for backorder".

2. Yes this should be the case

3. Yes, agreed.

I concur that the backorder message appears on the shopping cart, does NOT appear on the minicart, but DOES appear in checkout. It does not appear on the product page add to cart confirmation message. I had to write these bits for my store.

**Comment by seansan** [comment]
Ok just to be clear to magento2 team

Many messages and situations are not correct and need to be fixed. Correct?

**Comment by gwharton** [comment]
Yes. It's not that they are broken, they just don't exist.

**Comment by seansan** [comment]
Haha I call this a broken process

Sorry. Closed loop processes. Only
Logical and complete.
